### PR TITLE
tfel: add v3.0.13 -> v4.2.0

### PR DESCRIPTION
**Pull request by thelfer**
….1.13 and 3.0.13

Also fix Issue https://github.com/thelfer/tfel/issues/536 by adding `LD_LIBRARY_PATH` to the runtime environment, which is required for `MFront` to work properly

### CODE REVIEW

File: var/spack/repos/builtin/packages/tfel/package.py [hunk 1]
```diff
@@ -194,6 +202,9 @@ def cmake_args(self):
 
         return args
 
+    def setup_run_environment(self, env):
+        env.append_path("LD_LIBRARY_PATH", self.prefix.lib)
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/tfel/package.py:206]: Why was this added?

**Review by alalazo — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/tfel/package.py (hunk 1 above)
Comment (thelfer) [reply var/spack/repos/builtin/packages/tfel/package.py:206]: The TFEL project provides several shared libraries and the MFront code generator.

Typical usage of MFront is to generate C++ files and compile them into shared libraries linked to the shared libraries of the `TFEL` project. We don't use RPATH when linking, so setting `LD_LIBRARY_PATH` is required.

Hope to be clear enough !

Comment (alalazo) [reply var/spack/repos/builtin/packages/tfel/package.py:206]: So, if I understand correctly:
1. MFront works correctly (meaning, using RPATH), because it's compiled by Spack using its wrappers
2. Code generated by MFront does not, because internally the tool doesn't use RPATH

Can we patch MFront instead, to add relevant RPATHs? That would be similar to what we do with compilers like `gcc` when we write spec files.

Comment (thelfer) [reply var/spack/repos/builtin/packages/tfel/package.py:206]: My bad, I was not clear enough. First of all, MFront works indeed correctly.

However, when a user of MFront generates a new shared library, this new shared library is linked with libraries rom the TFEL library. This works fine.

The shared library that he has generated is however unusable since the TFEL libraries are not listed in `LD_LIBRARY_PATH`.

A typical usage of  MFront is:

```bash
$ mfront --obuild --interface=abaqus mybehaviour.mfront
The following libraries have been generated:
- libAbaqusBehaviour.so: MYBEHAVIOUR_TRIDIMENSIONAL
```

`libAbaqusBehaviour.so` is indeed linked with `libTFELMaterial.so` for instance. If you currently try to use `libAbaqusBehaviour.so`, the linker complains about  `libTFELMaterial.so` not being available. This is the reason why `LD_LIBRARY_PATH` must be set.

> Can we patch MFront instead, to add relevant RPATHs? 

There are two main reasons (possibly bad ones) why I would not do that:

- MFront supports several of compilers (`gcc`, `clang`, `icpx`, `cl.exe`, `nvcc`) on several platforms (`Linux`, `MacOs`, `Windows`, `FreeBSD`, `OpenSolaris`, `Haiku`, etc...) and two build systems (`make` and `cmake`). In MFront generates either a Makefile file or a CMakeLists.txt file to generate the shared libraries. Adding RPATHs would require to check that we don't break anything in all those configurations.
- We move our binaries in our CI when deploying those shared libraries (both the TFEL libraries and the generated libraries. This would need us to remove the RPATHs which is cumbersome.

Comment (alalazo) [reply var/spack/repos/builtin/packages/tfel/package.py:206]: > My bad, I was not clear enough.

I think you were clear, and we are saying the same thing :slightly_smiling_face: Let me think a bit on the `LD_LIBRARY_PATH` solution. I understand the issues, but otoh it's not the idiomatic way of packaging in Spack. 

If I don't get back in a few days, please ping me or DM me in Slack.

Comment (thelfer) [reply var/spack/repos/builtin/packages/tfel/package.py:206]: Perfect ! Any better solution would be welcomed.

BTW, updating LD_LIBRARY_PATH is done in a few other `SPACK` packages:

```bash
$ grep -l 'append_path("LD_LIBRARY_PATH"' $(find . -name "package.py")
./packages/aws-ofi-nccl/package.py
./packages/faiss/package.py
./packages/dtf/package.py
./packages/simmetrix-simmodsuite/package.py
./packages/cuda/package.py
./packages/acfl/package.py
./packages/likwid/package.py
./packages/hdf5/package.py
./packages/tau/package.py
```

Comment (thelfer) [reply var/spack/repos/builtin/packages/tfel/package.py:206]: Hi @alalazo. A little ping as requested. Did you think of a better way to handle this ?

Comment (alalazo) [reply var/spack/repos/builtin/packages/tfel/package.py:206]: I'll merge this as is, for the time being, but certainly this is not an idiom we want to promote for wide use :slightly_smiling_face: Also, if MFront could add RPATHs that would solve all the issues with Spack, and wouldn't require setting `LD_LIBRARY_PATH`.

**Review by alalazo — APPROVED** [review]

**Comment by thelfer** [comment]
Thanks @alalazo
